mutation_writer/multishard_writer.cc

```diff
@@ -131,16 +132,26 @@ future<stop_iteration> multishard_writer::handle_mutation_fragment(mutation_frag
     auto f = make_ready_future<>();
     if (mf.is_partition_start()) {
         _consumed_partitions++;
-        if (unsigned shard = shard_for_mf(mf); shard != _current_shard) {
-            _current_shard = shard;
-            if (!bool(_shard_writers[shard])) {
-                f = make_shard_writer(shard);
+        if (auto shards = shard_for_mf(mf); shards != _current_shards) {
+            _current_shards = shards;
+            // FIXME: Keep effective_replication_map ptr as long as we use those shards.
```

Comment:
I "fixed" it by saying that users of multishard writer should keep the erm alive, the one used to obtain the sharder which they pass to the multishard writer. They currently do.

Reply:
ok, please adjust the comment. It's not that we need to fix something to keep erm alive, it's that we need to fix the whole thing to avoid the requirement.

Reply:
This comment was removed. Added doc in topology-over-raft.md:
```
Do I have to hold to effective_replication_map_ptr around writing on the replica side?

Yes. Topology coordinator needs to wait for all writes to a tablet replica before cleaning it up to uphold [INV-TABL-1].
Holding on to effective_replication_map_ptr on the coordinator side is not enough since the coordinator may
already time-out or restart.

Alternatively, if the writes are done on behalf of a topology operation (e.g. tablet migration), it's enough to use
the topology guard mechanism, and hold the guard around writes instead of effective_replication_map_ptr.

```

and on top of distribute_reader_and_consume_on_shards():
```
/// The caller is responsible for ensuring that effective_replication_map_ptr used to obtain the sharder
/// is alive until the returned future is ready. Alternatively, if auto_refreshing_sharder is used,
/// the topology_guard associated with operation must be used and checked by the consumer.
future<uint64_t> distribute_reader_and_consume_on_shards(schema_ptr s,
```